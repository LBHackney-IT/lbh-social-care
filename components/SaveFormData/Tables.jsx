import TimesCircleIcon from 'components/Icons/TimesCircle';
import Link from 'next/link';

export const DetailedTable = ({ tableHeader, data, deleteForm }) => (
  <table className="govuk-table">
    <thead className="govuk-table__head">
      <tr className="govuk-table__row">
        {tableHeader.map((text) => (
          <th key={text} className="govuk-table__header">
            {text}
          </th>
        ))}
      </tr>
    </thead>
    <tbody className="govuk-table__body">
      {data.map((cell) => (
        <tr key={'key'} className="govuk-table__row">
          <td className="govuk-table__cell">{`#${cell.data.id}`} </td>
          <td className="govuk-table__cell">
            {`${cell.firstName} ${cell.lastName}`}{' '}
          </td>
          <td className="govuk-table__cell">
            {cell.dateOfBirth &&
              new Date(cell.dateOfBirth).toLocaleDateString('en-GB')}{' '}
          </td>
          <td className="govuk-table__cell">{cell.title} </td>
          <td className="govuk-table__cell"> {cell.timeStamp}</td>
          <td className="govuk-table__cell">
            {
              <Link href={`${cell.step}&continueForm=true`}>
                <a className="govuk-link">Complete</a>
              </Link>
            }
          </td>
          <td
            role="button"
            className="govuk-table__cell"
            onClick={() => deleteForm(cell.formPath)}
          >
            <TimesCircleIcon color="danger" />
          </td>
        </tr>
      ))}
    </tbody>
  </table>
);

export const StandardTable = ({ tableHeader, data, deleteForm }) => (
  <table className="govuk-table">
    <thead className="govuk-table__head">
      <tr className="govuk-table__row">
        {tableHeader.map((text) => (
          <th key={text} className="govuk-table__header">
            {text}
          </th>
        ))}
      </tr>
    </thead>
    <tbody className="govuk-table__body">
      {data.map((cell) => (
        <tr key={cell.formPath} className="govuk-table__row">
          <td className="govuk-table__cell">{cell.title} </td>
          <td className="govuk-table__cell"> {cell.timeStamp}</td>
          <td className="govuk-table__cell">
            {
              <Link href={`${cell.step}?continueForm=true`}>
                <a className="govuk-link">Complete</a>
              </Link>
            }
          </td>
          <td
            role="button"
            className="govuk-table__cell"
            onClick={() => deleteForm(cell.formPath)}
          >
            <TimesCircleIcon color="danger" />
          </td>
        </tr>
      ))}
    </tbody>
  </table>
);
