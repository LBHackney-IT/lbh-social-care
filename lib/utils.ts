import { Form, Step, Field } from '../data/flexibleForms/forms.types';
import { Resident } from 'types';

export interface Theme {
  name: string;
  steps: Step[];
}

type InitialValue = null | string | string[] | InitialValues[];

export interface InitialValues {
  [key: string]: InitialValue;
}

/** Take a form's steps and organise them by theme for displaying in a task list */
export const groupByTheme = (form: Form): Theme[] =>
  form.steps.reduce((groups: Theme[], step) => {
    const name = step.theme;
    if (name && !groups.find((group) => group.name === name)) {
      groups.push({
        name: name,
        steps: [],
      });
    }
    const group = groups.find((group) => group.name === name);
    group?.steps.push(step);
    return groups;
  }, []);

/** Shorten a long string to a given number of words for displaying in previews */
export const truncate = (str: string, noWords: number): string => {
  if (str.split(' ').length > noWords) {
    return str.split(' ').splice(0, noWords).join(' ') + '...';
  } else {
    return str;
  }
};

const initiallyNull = new Set(['file']);
const initiallyFirstChoice = new Set(['select']);
const initiallyArray = new Set(['checkboxes', 'repeater']);

/** Generate flexible initial values for a flexible schema */
export const generateInitialValues = (
  fields: Field[],
  person?: Resident
): InitialValues => {
  const initialValues: InitialValues = {};

  fields.map((field) => {
    if (field.type === 'repeaterGroup') {
      initialValues[field.id] = [
        generateInitialValues(field.subfields || [], person),
      ];
    } else if (initiallyArray.has(field.type)) {
      initialValues[field.id] = [];
    } else if (initiallyNull.has(field.type)) {
      initialValues[field.id] = null;
    } else if (initiallyFirstChoice.has(field.type)) {
      initialValues[field.id] = String(
        (person && field.prefill && person[field.prefill]) ||
          (field.choices && field.choices[0].value)
      );
    } else {
      initialValues[field.id] = String(
        (person && field.prefill && person[field.prefill]) || ''
      );
    }
  });
  return initialValues;
};

/** Push an element into an array, without duplicates */
export const pushUnique = <T>(array: T[], newElement: T): T[] =>
  Array.from(new Set(array).add(newElement));
