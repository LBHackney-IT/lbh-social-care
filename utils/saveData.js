export const windowGlobal = typeof window !== 'undefined' && window;

export const saveData = (
  formPath,
  data,
  title,
  step,
  includesDetails,
  personDetails
) => {
  const timeStamp = new Date(Date.now()).toLocaleString().split(',')[0];
  const savedData = {
    step,
    data,
    title,
    timeStamp,
    formPath,
    includesDetails,
    ...personDetails,
  };
  try {
    windowGlobal.localStorage.setItem(formPath, JSON.stringify(savedData));
  } catch {
    deleteData(formPath);
  }
};

export const getData = (formPath) => {
  try {
    return JSON.parse(windowGlobal.localStorage.getItem(formPath));
  } catch {
    deleteData(formPath);
  }
};

export const getDataIncludes = (includes) => {
  if (windowGlobal.localStorage) {
    try {
      const data = Object.fromEntries(
        Object.entries(windowGlobal.localStorage)
          .filter(([key]) => key.includes(includes))
          .map(([key, value]) => [key, JSON.parse(value)])
      );
      return Object.keys(data).length > 0 ? data : null;
    } catch {
      windowGlobal.localStorage.clear();
      return null;
    }
  }
};

export const deleteData = (formPath) => {
  windowGlobal.localStorage.removeItem(formPath);
};
