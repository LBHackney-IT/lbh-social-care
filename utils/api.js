import { getQueryString } from 'utils/urls';

export const getInfiniteKey = (baseUrl, key, params) => (
  pageIndex,
  previousPageData
) => {
  if (previousPageData && !previousPageData[key]) return null;
  if (pageIndex === 0)
    return `${baseUrl}${params ? `?${getQueryString(params)}` : ''}`;
  return `${baseUrl}?${getQueryString({
    ...params,
    cursor: previousPageData.nextCursor,
  })}`;
};
