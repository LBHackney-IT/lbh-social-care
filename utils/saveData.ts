import { isBrowser } from 'utils/ssr';
import { Resident } from 'types';

export const SAVE_KEY = 'social-care-forms_v2';

export interface SavedData {
  data: Record<string, unknown>;
  title: string;
  formPath: string;
  timestamp: string;
  step: string;
  personDetails?: Resident;
}

type DataToSave = Omit<SavedData, 'timestamp'>;

export const saveData = (data: DataToSave): void => {
  try {
    const savedForms = localStorage.getItem(SAVE_KEY) || '{}';
    savedForms &&
      localStorage.setItem(
        SAVE_KEY,
        JSON.stringify({
          ...JSON.parse(savedForms),
          [data.formPath]: {
            ...data,
            timestamp: new Date().toLocaleDateString('en-GB'),
          },
        })
      );
  } catch {
    localStorage.removeItem(SAVE_KEY);
  }
};

export const getFormData = (formPath: string): SavedData | undefined => {
  if (isBrowser()) {
    try {
      const savedForms = localStorage.getItem(SAVE_KEY);
      return savedForms && JSON.parse(savedForms)?.[formPath];
    } catch {
      localStorage.removeItem(SAVE_KEY);
    }
  }
};

export const getData = (): Record<string, SavedData> | undefined => {
  if (isBrowser()) {
    try {
      const savedForms = localStorage.getItem(SAVE_KEY);
      return savedForms && JSON.parse(savedForms);
    } catch {
      localStorage.removeItem(SAVE_KEY);
    }
  }
};

export const deleteData = (formPath: string): void => {
  try {
    const savedForms = JSON.parse(localStorage.getItem(SAVE_KEY) || '');
    if (savedForms) delete savedForms[formPath];
    localStorage.setItem(SAVE_KEY, JSON.stringify(savedForms));
  } catch {
    localStorage.removeItem(SAVE_KEY);
  }
};
