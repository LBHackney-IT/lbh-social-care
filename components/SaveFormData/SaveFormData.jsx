import { useState, useEffect } from 'react';

import { getData, deleteData } from 'utils/saveData';
import { DetailedTable, StandardTable } from './Tables';

export const SavedForms = () => {
  const [savedForms, setSavedForms] = useState();
  useEffect(() => {
    setSavedForms(getData());
  }, []);
  const deleteForm = (path) => {
    deleteData(path);
    setSavedForms(getData());
  };
  if (!savedForms || Object.keys(savedForms)?.length === 0) {
    return (
      <p role="label" className="govuk-fieldset__legend--m gov-weight-lighter">
        You don't have any incomplete form, well done!
      </p>
    );
  }
  const detailHeader = [
    'Person ID',
    'Client Name',
    'Date of birth',
    'Form type',
    'Last saved',
    'Actions',
    '',
  ];
  const standardHeader = ['Form type', 'Last saved', 'Actions', ''];
  const sortData = Object.values(savedForms);
  const formQty = sortData.length;
  const detailData = sortData.filter((item) => item.includesDetails);
  const standardData = sortData.filter((item) => !item.includesDetails);
  return (
    <>
      <p role="label" className="govuk-fieldset__legend--s gov-weight-lighter">
        {`Displaying ${formQty} unfinished ${formQty > 1 ? 'forms' : 'form'}`}{' '}
      </p>
      {standardData.length ? (
        <StandardTable
          tableHeader={standardHeader}
          data={standardData}
          deleteForm={deleteForm}
        />
      ) : null}
      {detailData.length ? (
        <DetailedTable
          tableHeader={detailHeader}
          data={detailData}
          deleteForm={deleteForm}
        />
      ) : null}
    </>
  );
};
