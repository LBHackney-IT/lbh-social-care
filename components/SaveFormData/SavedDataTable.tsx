import Link from 'next/link';

import TimesCircleIcon from 'components/Icons/TimesCircle';
import { SavedData } from 'utils/saveData';

interface Props {
  tableHeader: Array<string>;
  savedData: Array<SavedData>;
  deleteForm: (arg0: string) => void;
}

const SavedDataTable = ({
  tableHeader,
  savedData,
  deleteForm,
}: Props): React.ReactElement => (
  <table className="govuk-table">
    <thead className="govuk-table__head">
      <tr className="govuk-table__row">
        {tableHeader.map((text) => (
          <th key={text} className="govuk-table__header">
            {text}
          </th>
        ))}
      </tr>
    </thead>
    <tbody className="govuk-table__body">
      {savedData.map((record) => (
        <tr key={'key'} className="govuk-table__row">
          {record.personDetails && (
            <>
              <td className="govuk-table__cell">
                {`#${record.personDetails.id}`}{' '}
              </td>
              <td className="govuk-table__cell">
                {`${record.personDetails.firstName} ${record.personDetails.lastName}`}{' '}
              </td>
              <td className="govuk-table__cell">
                {record.personDetails.dateOfBirth &&
                  new Date(record.personDetails.dateOfBirth).toLocaleDateString(
                    'en-GB'
                  )}{' '}
              </td>
            </>
          )}
          <td className="govuk-table__cell">{record.title} </td>
          <td className="govuk-table__cell"> {record.timestamp}</td>
          <td className="govuk-table__cell">
            {
              <Link
                href={`${record.step}${
                  record.step.indexOf('?') === -1 ? '?' : '&'
                }continueForm=true`}
              >
                <a className="govuk-link">Complete</a>
              </Link>
            }
          </td>
          <td
            role="button"
            className="govuk-table__cell"
            onClick={() => deleteForm(record.formPath)}
          >
            <TimesCircleIcon color="danger" />
          </td>
        </tr>
      ))}
    </tbody>
  </table>
);

export default SavedDataTable;
