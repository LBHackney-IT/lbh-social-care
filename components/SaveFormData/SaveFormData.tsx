import { useState, useEffect } from 'react';

import { getData, deleteData, SavedData } from 'utils/saveData';
import SavedDataTable from './SavedDataTable';

export const SavedForms = (): React.ReactElement => {
  const [savedForms, setSavedForms] = useState<Record<string, SavedData>>();
  useEffect(() => {
    setSavedForms(getData());
  }, []);
  const deleteForm = (path: string) => {
    deleteData(path);
    setSavedForms(getData());
  };
  if (!savedForms || Object.keys(savedForms)?.length === 0) {
    return (
      <p role="label" className="govuk-fieldset__legend--m gov-weight-lighter">
        You don&apos;t have any incomplete form, well done!
      </p>
    );
  }
  const detailHeader = ['Person ID', 'Client Name', 'Date of birth'];
  const standardHeader = ['Form type', 'Last saved', 'Actions', ''];
  const sortData = Object.values(savedForms);
  const formQty = sortData.length;
  const detailData = sortData.filter((item) => Boolean(item.personDetails));
  const standardData = sortData.filter((item) => !item.personDetails);
  return (
    <>
      <p role="label" className="govuk-fieldset__legend--s gov-weight-lighter">
        {`Displaying ${formQty} unfinished ${formQty > 1 ? 'forms' : 'form'}`}{' '}
      </p>
      {standardData.length > 0 && (
        <SavedDataTable
          tableHeader={standardHeader}
          savedData={standardData}
          deleteForm={deleteForm}
        />
      )}
      {detailData.length > 0 && (
        <SavedDataTable
          tableHeader={[...detailHeader, ...standardHeader]}
          savedData={detailData}
          deleteForm={deleteForm}
        />
      )}
    </>
  );
};
