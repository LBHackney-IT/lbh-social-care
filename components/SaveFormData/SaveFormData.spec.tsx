import { render } from '@testing-library/react';

import { SavedForms } from './SaveFormData';
import * as saveData from 'utils/saveData';

jest.mock('utils/saveData', () => ({
  getData: jest.fn(),
}));

describe('SaveFormData component', () => {
  it('should render both types of forms', async () => {
    jest.spyOn(saveData, 'getData').mockImplementationOnce(() => ({
      first: {
        formPath: '/form/foo-bar/',
        timeStamp: '22/12/2020',
        title: 'Foo Bar',
        step: 'foo',
      } as saveData.BasicData,
      second: {
        dateOfBirth: '1984-02-12T00:00:00.0000000',
        firstName: 'Foo',
        formPath: '/form/bar-foo/',
        gender: 'M',
        includesDetails: true,
        lastName: 'Bar',
        mosaicId: '1234',
        nhsNumber: '123',
        step: 'foo',
        timeStamp: '22/12/2020',
        title: 'Foo Bar',
        data: { id: '12345' },
      },
    }));
    const { getByRole, asFragment } = render(<SavedForms />);
    expect(asFragment()).toMatchSnapshot();
    expect(getByRole('label')).toHaveTextContent(
      'Displaying 2 unfinished form'
    );
  });

  it('should display no forms in progress message', () => {
    const { getByRole } = render(<SavedForms />);
    expect(getByRole('label')).toHaveTextContent(
      `You don't have any incomplete form, well done!`
    );
  });
});
