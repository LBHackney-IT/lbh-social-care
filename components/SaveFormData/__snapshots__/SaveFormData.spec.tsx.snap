// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SaveFormData component should render both types of forms 1`] = `
<DocumentFragment>
  <p
    class="lbh-body"
  >
    You have 2 unfinished forms 
  </p>
  <table
    class="govuk-table"
  >
    <thead
      class="govuk-table__head"
    >
      <tr
        class="govuk-table__row"
      >
        <th
          class="govuk-table__header"
        >
          Form type
        </th>
        <th
          class="govuk-table__header"
        >
          Last saved
        </th>
        <th
          class="govuk-table__header"
        >
          Actions
        </th>
        <th
          class="govuk-table__header"
        />
      </tr>
    </thead>
    <tbody
      class="govuk-table__body"
    >
      <tr
        class="govuk-table__row"
      >
        <td
          class="govuk-table__cell"
        >
          Foo Bar 
        </td>
        <td
          class="govuk-table__cell"
        >
           22/12/2020
        </td>
        <td
          class="govuk-table__cell"
        >
          <a
            class="govuk-link"
            href="/foo?continueForm=true"
          >
            Complete
          </a>
        </td>
        <td
          class="govuk-table__cell"
          role="button"
        >
          <svg
            class="cancel-icon"
            height="512pt"
            viewBox="0 0 512 512"
            width="512pt"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m256 0c-141.164062 0-256 114.835938-256 256s114.835938 256 256 256 256-114.835938 256-256-114.835938-256-256-256zm0 0"
              fill="#d4351c"
            />
            <path
              d="m350.273438 320.105469c8.339843 8.34375 8.339843 21.824219 0 30.167969-4.160157 4.160156-9.621094 6.25-15.085938 6.25-5.460938 0-10.921875-2.089844-15.082031-6.25l-64.105469-64.109376-64.105469 64.109376c-4.160156 4.160156-9.621093 6.25-15.082031 6.25-5.464844 0-10.925781-2.089844-15.085938-6.25-8.339843-8.34375-8.339843-21.824219 0-30.167969l64.109376-64.105469-64.109376-64.105469c-8.339843-8.34375-8.339843-21.824219 0-30.167969 8.34375-8.339843 21.824219-8.339843 30.167969 0l64.105469 64.109376 64.105469-64.109376c8.34375-8.339843 21.824219-8.339843 30.167969 0 8.339843 8.34375 8.339843 21.824219 0 30.167969l-64.109376 64.105469zm0 0"
              fill="#fafafa"
            />
          </svg>
        </td>
      </tr>
    </tbody>
  </table>
  <table
    class="govuk-table"
  >
    <thead
      class="govuk-table__head"
    >
      <tr
        class="govuk-table__row"
      >
        <th
          class="govuk-table__header"
        >
          Person ID
        </th>
        <th
          class="govuk-table__header"
        >
          Client Name
        </th>
        <th
          class="govuk-table__header"
        >
          Date of birth
        </th>
        <th
          class="govuk-table__header"
        >
          Form type
        </th>
        <th
          class="govuk-table__header"
        >
          Last saved
        </th>
        <th
          class="govuk-table__header"
        >
          Actions
        </th>
        <th
          class="govuk-table__header"
        />
      </tr>
    </thead>
    <tbody
      class="govuk-table__body"
    >
      <tr
        class="govuk-table__row"
      >
        <td
          class="govuk-table__cell"
        >
          #12345 
        </td>
        <td
          class="govuk-table__cell"
        >
          Foo Bar 
        </td>
        <td
          class="govuk-table__cell"
        >
          12/02/1984 
        </td>
        <td
          class="govuk-table__cell"
        >
          Foo Bar 
        </td>
        <td
          class="govuk-table__cell"
        >
           22/12/2020
        </td>
        <td
          class="govuk-table__cell"
        >
          <a
            class="govuk-link"
            href="/foo&continueForm=true"
          >
            Complete
          </a>
        </td>
        <td
          class="govuk-table__cell"
          role="button"
        >
          <svg
            class="cancel-icon"
            height="512pt"
            viewBox="0 0 512 512"
            width="512pt"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m256 0c-141.164062 0-256 114.835938-256 256s114.835938 256 256 256 256-114.835938 256-256-114.835938-256-256-256zm0 0"
              fill="#d4351c"
            />
            <path
              d="m350.273438 320.105469c8.339843 8.34375 8.339843 21.824219 0 30.167969-4.160157 4.160156-9.621094 6.25-15.085938 6.25-5.460938 0-10.921875-2.089844-15.082031-6.25l-64.105469-64.109376-64.105469 64.109376c-4.160156 4.160156-9.621093 6.25-15.082031 6.25-5.464844 0-10.925781-2.089844-15.085938-6.25-8.339843-8.34375-8.339843-21.824219 0-30.167969l64.109376-64.105469-64.109376-64.105469c-8.339843-8.34375-8.339843-21.824219 0-30.167969 8.34375-8.339843 21.824219-8.339843 30.167969 0l64.105469 64.109376 64.105469-64.109376c8.34375-8.339843 21.824219-8.339843 30.167969 0 8.339843 8.34375 8.339843 21.824219 0 30.167969l-64.109376 64.105469zm0 0"
              fill="#fafafa"
            />
          </svg>
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;
